src/audio/dp_queue.c

```diff
@@ -0,0 +1,367 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2021 Intel Corporation. All rights reserved.
+//
+
+#include <sof/audio/dp_queue.h>
+#include <sof/audio/sink_api_implementation.h>
+#include <sof/audio/source_api_implementation.h>
+#include <sof/audio/audio_stream.h>
+
+#include <rtos/alloc.h>
+#include <ipc/topology.h>
+
+struct dp_queue {
+	struct sof_source source_api;  /**< src api handler */
+	struct sof_sink sink_api;      /**< sink api handler */
+
+	uint32_t ibs;
+	uint32_t obs;
+	uint32_t flags;
+
+	size_t data_buffer_size;
+	uint8_t __sparse_cache *data_buffer;
+
+	uint32_t write_offset;
+	uint32_t read_offset;
+	uint32_t available_data; /* amount of data ready for immediate reading */
+	/* free buffer space
+	 * NOTE!
+	 *  - when dp queue is shared, available_data + free_space DOES NOT eq data_buffer_size
+	 *  - when dp queue is not shared available_data + free_space always == data_buffer_size
+	 */
+	uint32_t free_space;
+
+	bool hw_params_configured;
+	struct sof_audio_stream_params audio_stream_params;
+
+	struct k_spinlock lock;
+};
+
+static inline bool dp_queue_is_shared(struct dp_queue *dp_queue)
+{
+	return !!(dp_queue->flags & DP_QUEUE_MODE_SHARED);
+}
+
+static inline uint8_t __sparse_cache *dp_queue_buffer_end(struct dp_queue *dp_queue)
+{
+	return dp_queue->data_buffer + dp_queue->data_buffer_size;
+}
+
+struct dp_queue *dp_queue_get_from_sink(struct sof_sink *sink)
+{
+	return container_of(sink, struct dp_queue, sink_api);
+}
+
+struct dp_queue *dp_queue_get_from_source(struct sof_source *source)
+{
+	return container_of(source, struct dp_queue, source_api);
+}
```

Comment:
These need to be static inline in the header - I dont think we can say API partitioning (i.e. what is public/private) is higher priority than MCPS.

Reply:
We need to double check if clang linker is able to inline such functions (I know there are "out there" linkers smart enough). We'll be able to eat a cake and have it
OR we can simply go to C++ ;)

Reply:
unfortunately they're needed outside, till Pipeline2.0 is ready. Now there's a need for some tricks in module adapter to make connection to existing design